### PR TITLE
Added basic file logger

### LINKED ISSUE #1803: Add a basic logger which does not require ports
## 🚀 Feature

It would be great to have a basic logger that does not require opening ports.

### Motivation

At this time, all loggers supported by Lightning requires to use ports. There are some environments where I don't have total control and I cannot open ports. It would be great to implement a logger that could dump the metrics during training, and easily visualize their evolution, without the need to forward a port.

### Pitch

As proposed by @justusschock in the slack channel, one could add a basic logger `FileLogger` that saves metrics into a file (like json) and plots them on a regular basis.

Thank you for the great package !

### PR DESCRIPTION
**Pull request by xmotli02**
## What does this PR do?

This PR implements basic file logger as discussed in #1803 

Fixes #1803 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: pytorch_lightning/loggers/file_logger.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
+"""
```
Comment (justusschock) [inline pytorch_lightning/loggers/file_logger.py:1]: File Logger may be a bit misleading, since tensor board generates a file as well :) Maybe rename this file to `csv.py` and the Logger to `CSVLogger`?

File: pytorch_lightning/loggers/file_logger.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+File logger
+-----------
+"""
```
Comment (justusschock) [inline pytorch_lightning/loggers/file_logger.py:4]: Some description here?

File: pytorch_lightning/loggers/file_logger.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+from typing import Optional, Dict, Any, Union
+
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning.core.saving import save_hparams_to_yaml
+from pytorch_lightning.loggers.base import LightningLoggerBase
+from pytorch_lightning.utilities.distributed import rank_zero_only
+
+
+class ExperimentWriter(object):
```
Comment (justusschock) [inline pytorch_lightning/loggers/file_logger.py:20]: docstrings :)

File: pytorch_lightning/loggers/file_logger.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    NAME_HPARAMS_FILE = 'hparams.yaml'
+    NAME_METRICS_FILE = 'metrics.csv'
+
+    def __init__(self, log_dir):
```
Comment (justusschock) [inline pytorch_lightning/loggers/file_logger.py:24]: typing :)

File: pytorch_lightning/loggers/file_logger.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+        Example::
+
+            self.logger.experiment.some_experiment_writer_function()
+
+        """
+        if self._experiment is not None:
+            return self._experiment
+
+        os.makedirs(self.root_dir, exist_ok=True)
```
Comment (justusschock) [inline pytorch_lightning/loggers/file_logger.py:133]: maybe we should warn the user here that potential logged values will be overwritten?

**Comment by Borda** [comment]
> This looks really nice  Agree that this should be changed to `CSVLogger`. It would also be good to add the `save_dir` property to the logger so that checkpoints are saved in the right place (this may just mean renaming the `root_dir` property?)

I am not very convinced about CSV format for logging, then you would have plant hard time with any separator you use...

**Comment by ethanwharris** [comment]
> > This looks really nice  Agree that this should be changed to `CSVLogger`. It would also be good to add the `save_dir` property to the logger so that checkpoints are saved in the right place (this may just mean renaming the `root_dir` property?)
> 
> I am not very convinced about CSV format for logging, then you would have plant hard time with any separator you use...

@Borda Whoops, should have said 'should be **renamed** to `CSVLogger`' since that's what it already does (but called `FileLogger` currently). Presumably if people want more options (e.g. different delimiters, JSON logging, etc.) then these can be added later on?

File: pytorch_lightning/loggers/file_logger.py (hunk 4 above)
Comment (xmotli02) [reply pytorch_lightning/loggers/file_logger.py:24]: Hi, I'm not sure what you mean by "typing"

File: pytorch_lightning/loggers/file_logger.py (hunk 5 above)
Comment (xmotli02) [reply pytorch_lightning/loggers/file_logger.py:133]: In that case, should the exception be raised or just warn the user and overwrite existing files?

Comment (xmotli02) [reply pytorch_lightning/loggers/file_logger.py]: Well, I left this line unchanged. This way it is consistent with the `TensorboardLogger`, see [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/tensorboard.py#L106). If the experiment `root_dir` folder exists, it's content will not be deleted. 

The `root_dir` folder contains subfolders with experiment runs (versions), where the actual metrics and hparams are saved. 
Directory structure:
```
root_dir/
  version_1/
    hparams.yaml
    metrics.csv
  version_2/
  ...
```

If the user does not select the experiment version, next available version is chosen - a new subfolder is created. However, if the user selects an already existing version, the content will be overwritten. Therefore I added a warning to `ExperimentWriter` for that particular case, see [line](https://github.com/PyTorchLightning/pytorch-lightning/blob/c6fc5d3fc43d00d2b75940d1b94b129420778c07/pytorch_lightning/loggers/csv.py#L44).

File: pytorch_lightning/loggers/file_logger.py (hunk 4 above)
Comment (justusschock) [reply pytorch_lightning/loggers/file_logger.py]: Adding type signatures to arguments and return types. for your init it would be `def __init__(self, log_dir: Union[str, path lib.Path]) -> None:`

**Comment by xmotli02** [comment]
> 
> 
> This looks really nice 😃 Agree that this should be ~changed~ _renamed_ to `CSVLogger`. It would also be good to add the `save_dir` property to the logger so that checkpoints are saved in the right place (this may just mean renaming the `root_dir` property?)

Added `save_dir` property, so that it is consistent with the TensorboardLogger implementation.

**Comment by xmotli02** [comment]
> 
> 
> This looks very similar to Test-Tube logger, mind clarify what is the difference?

@Borda Yes, it's true it was inspired by the Test-Tube logger as suggested in [this](https://github.com/PyTorchLightning/pytorch-lightning/issues/1803#issuecomment-627486896) comment. The difference is that this logger has no 3rd party deps (pandas).

**Comment by Borda** [comment]
well, would it be easier to make `test-tube` lighter? @williamFalcon

File: pytorch_lightning/loggers/csv.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+        existing_versions = []
+        for d in os.listdir(root_dir):
+            if os.path.isdir(os.path.join(root_dir, d)) and d.startswith("version_"):
+                existing_versions.append(int(d.split("_")[1]))
+
+        if len(existing_versions) == 0:
+            return 0
+
+        return max(existing_versions) + 1
```
Comment (awaelchli) [inline pytorch_lightning/loggers/csv.py:204]: this is duplicated code from the tensorboard logger. can we please pull it out into a helper function?

Comment (awaelchli) [reply pytorch_lightning/loggers/csv.py:204]: I'm surprised the codefactor tool did not catch this :)

Comment (Borda) [reply pytorch_lightning/loggers/csv.py:204]: I guess it focus on diff sections for PRs

**Comment by williamFalcon** [comment]
@Borda can we finish for 0.9?

File: pytorch_lightning/loggers/csv.py (hunk 6 above)
Comment (xmotli02) [reply pytorch_lightning/loggers/csv.py:204]: Ou, it is my fault. 

Should I pull it out from the TensorBoardLogger into a separate file? Or leave it in the TensorBoardLogger and include it from it in CSVLogger?

File: pytorch_lightning/loggers/csv.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        """Record metrics"""
+        def _handle_value(value):
+            if isinstance(value, torch.Tensor):
+                return value.item()
+            return value
+
+        if step is None:
+            step = len(self.metrics)
+
+        new_row = dict.fromkeys(self.metrics_keys)
```
Comment (Borda) [inline pytorch_lightning/loggers/csv.py:66]: I do not think that is needed as it maintains only all future rows are complete but some may be missing in past...
I would rather just save as it is and do write a padding method for all rows according to the max

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,6 +9,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- Added CSVLogger ([#2721](https://github.com/PyTorchLightning/pytorch-lightning/pull/2721))
```
Comment (Borda) [inline CHANGELOG.md:12]:
```suggestion
- Added basi `CSVLogger` ([#2721](https://github.com/PyTorchLightning/pytorch-lightning/pull/2721))
```

File: pytorch_lightning/loggers/csv_logs.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+----------
+
+CSV logger for basic experiment logging that does not require opening ports
+
+"""
+import io
+import os
+import csv
+import torch
+
```
Comment (Borda) [inline pytorch_lightning/loggers/csv_logs.py:12]:
```suggestion
```

File: pytorch_lightning/loggers/csv_logs.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+from argparse import Namespace
+from typing import Optional, Dict, Any, Union
+
```
Comment (Borda) [inline pytorch_lightning/loggers/csv_logs.py:15]:
```suggestion
```

File: pytorch_lightning/loggers/csv_logs.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+    def __init__(self, log_dir: str) -> None:
+        self.hparams = {}
+        self.metrics = []
+
+        self.log_dir = log_dir
+        if os.path.exists(self.log_dir):
+            rank_zero_warn(
+                f"Experiment logs directory {self.log_dir} exists and is not empty. "
+                "Previous log files in this directory will be deleted when the new ones are saved!"
```
Comment (Borda) [inline pytorch_lightning/loggers/csv_logs.py:45]:
```suggestion
                f"Experiment logs directory {self.log_dir} exists and is not empty."
                " Previous log files in this directory will be deleted when the new ones are saved!"
```

File: pytorch_lightning/loggers/csv_logs.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        self._experiment = None
+
+    @property
+    def root_dir(self) -> str:
+        """
+        Parent directory for all checkpoint subdirectories.
+        If the experiment name parameter is ``None`` or the empty string, no experiment subdirectory is used
+        and the checkpoint will be saved in "save_dir/version_dir"
+        """
+        if self.name is None or len(self.name) == 0:
```
Comment (Borda) [inline pytorch_lightning/loggers/csv_logs.py:124]:
```suggestion
        if not self.name:
```

File: pytorch_lightning/loggers/csv_logs.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+            return self._save_dir
+        return os.path.join(self._save_dir, self.name)
```
Comment (Borda) [inline pytorch_lightning/loggers/csv_logs.py:126]:
```suggestion
            return self.save_dir
        return os.path.join(self.save_dir, self.name)
```

File: pytorch_lightning/loggers/csv_logs.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+
+        Actual ExperimentWriter object. To use ExperimentWriter features in your
+        :class:`~pytorch_lightning.core.lightning.LightningModule` do the following.
+
+        Example::
+
+            self.logger.experiment.some_experiment_writer_function()
+
+        """
+        if self._experiment is not None:
```
Comment (Borda) [inline pytorch_lightning/loggers/csv_logs.py:156]:
```suggestion
        if self._experiment:
```

File: tests/loggers/test_all.py [hunk 15]
```diff
@@ -226,6 +232,7 @@ def on_batch_start(self, trainer, pl_module):
 @pytest.mark.skipif(platform.system() == "Windows", reason="Distributed training is not supported on Windows")
 @pytest.mark.parametrize("logger_class", [
     TensorBoardLogger,
+    # CSVLogger,  # todo
```
Comment (Borda) [inline tests/loggers/test_all.py:235]: any idea why this test is failing?
cc: @PyTorchLightning/core-contributors
